### PR TITLE
Allow formats to modify templates without requiring a mod

### PR DESCRIPTION
**Pull request by urkerab**
Changing Pokemon's stats is a popular OM idea; 350 Cup and Averagemons are two obvious users. This is normally done using a mod, but I understand that it would be preferable to avoid this, so I have come up with this idea. You would then put code something like this in the format:

	onBeforeFormeChange: function (template, target, source, effect) {
		if (source) return;
		if (Object.values(template.baseStats).reduce((x, y) => x + y) > 350) return;
		template = Object.assign({}, template);
		template.baseStats = Object.assign({}, template.baseStats);
		for (let statName in template.baseStats) template.baseStats[statName] *= 2;
		return template;
	},

Note that this doesn't (yet) affect the HP, as that is currently calculated before the first forme change. There are at least three potential options:

 - Ignore the HP, and get the format to recalculate it in the `onBegin` event
 - Calculate the HP after the first forme change, breaking existing tier shift code
 - Make forme changing recalculate the HP (move the code for Zygarde Complete)

### CODE REVIEW

**Comment by Zarel** [comment]
Perhaps `onGetBaseStats`?

**Comment by urkerab** [comment]
I guess we can just do the stats for now... although I vaguely remember a Gen 6 OM where mega evolving didn't change your ability and this could have been achieved without resorting to a mod by resetting the mega template to use the set's Ability.

**Comment by Zarel** [comment]
We can also calculate HP after the first forme change, that's fine with me, too.

**Comment by Zarel** [comment]
I would probably call this event `ModifyTemplate`.

**Comment by urkerab** [comment]
For comparison, commit 27a9dc7 shows what just modifying the stats would look like. The 350 Cup code would then look something like this:

		spreadModify: function (stats, set, template) {
			if (Object.values(stats).reduce((x, y) => x + y) <= 350) {
				for (let statName in stats) stats[statName] *= 2;
			}
			return this.spreadModify(stats, set);
		},

**Comment by Zarel** [comment]
I'm okay with modifying templates, but shadowing `spreadModify` is entirely wrong. Why not the event system you were doing before?

Also, @urkerab, your code is an infinite recursive loop? You're not calling `super.spreadModify` here, you're just calling yourself repeatedly.

**Comment by urkerab** [comment]
Back to modifying templates, and now calculating HP after the first form change.
The simpler change to Hidden Type would be just to copy the assignment to `baseStats` from `battle-engine.js`, but I thought this way was slightly neater, although it's annoying that Reflect Type is still a special case.

File: battle-engine.js [hunk 1]
```diff
@@ -697,18 +696,24 @@ class BattlePokemon {
...
 
 		this.types = template.types;
-		this.addedType = '';
+		this.addedType = template.addedType || '';
 		this.knownType = true;
 
-		if (!dontRecalculateStats) {
+		if (!source) {
 			let stats = this.battle.spreadModify(this.template.baseStats, this.set);
+			if (!this.baseStats) this.baseStats = stats;
```
Comment (urkerab) [inline battle-engine.js]: Slightly confusing I guess given line 719 but I need this in order to set the base hp.

**Comment by Zarel** [comment]
I approve.

**Comment by Zarel** [comment]
Looks great!

**Pull request merged**